### PR TITLE
Global notifications component

### PR DESCRIPTION
**Pull request by Hyperkid123**
Added notifications component. So far is using PF4 Alert component.

## How to initialize
1. You should use insights redux API

```JSX
import { notifications } from '.@red-hat-insights/insights-frontend-components/SmartComponents/Notifications/';
import { getRegistry } from '@red-hat-insights/insights-frontend-components/';

// somewhere where are you adding reducers to store
this.reducerRegistry = getRegistry()
this.reducerRegistry.register({ notifications })
```

### How to use
1. Place the NotificationsPortal component (probably near root component). It mist be inside store Provider
```JSX
import { NotificationsPortal } from '.@red-hat-insights/insights-frontend-components/SmartComponents/Notifications/';

<Provider store={...}>
{/**
* Somewhere inside provideer
*/}
    <NotificationsPortal />
</Provider>
```
2. Import redux actions that control notifications and use them (Note: for basic usage you will need just addNotification)
```JSX
import { addNotification, removeNotification, clearNotifications } from '.@red-hat-insights/insights-frontend-components/SmartComponents/Notifications/';

// Add actions to your component
const mapDispatchToProps = dispatch => bindActionCreators({
  addNotification,
  removeNotification,
  clearNotifications
}, dispatch)

connect(mapStateToProps, mapDispatchToProps)(YourComponent)

// somewehere in render
<button
  onClick={() => this.props.addNotification({
    variant: 'success',
    title: 'Notification title',
    description: 'notification description'
  })}
>
  Add notification
</button>
```

## Define notification object
```javascript
{
  title: string,
  variant: ['success', 'danger', 'info', 'warning'],
  description: string?
  dismissable: bool? // if is false, notification will disappear automatically after 5s, otherwise will show close button,
  dismissDelay: number? // time period after which alert disappears in ms. Default value is 5000
}
```

## Demo

screencast-localhost-4005-2018 10 25-13-23-56

### CODE REVIEW

**Comment by Hyperkid123** [comment]
cc @karelhala

File: src/SmartComponents/Notifications/Notification.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                    <Button
+                        aria-label="close-notification"
+                        variant="plain"
+                        onClick={ this.handleDismiss }
+                    >
+                        <CloseIcon/>
+                    </Button> : null
+                }
+            >
+                { description && <p>{ description }</p> }
```
Comment (karelhala) [inline src/SmartComponents/Notifications/Notification.js]: No need to wrap it with `p` element, we should rather file a new PR addressing this change in pf-react.

File: src/SmartComponents/Notifications/NotificationPortal.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { createPortal } from 'react-dom';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import { removeNotification } from '../../redux/actions/notifications';
+import Notification from './Notification';
+import './notifications.scss';
+
+const Portal = ({ notifications, removeNotification, rootId }) =>
+    !notifications || Array.isArray(notifications) && notifications.length === 0
```
Comment (karelhala) [inline src/SmartComponents/Notifications/NotificationPortal.js]: Please wrap it with parenthesis so it's easier to read.
```
(!notifications || Array.isArray(notifications)) && notifications.length === 0
```

File: src/redux/action-types.js [hunk 3]
```diff
@@ -30,3 +30,11 @@ export const SELECT_ENTITY = 'SELECT_ENTITY';
 export const CHANGE_SORT = 'CHANGE_SORT';
 export const FILTER_ENTITIES = 'FILTER_ENTITIES';
 export const APPLICATION_SELECTED = 'APPLICATION_SELECTED';
+
+/**
+ * Notifications actions
+ */
+const notificationsPrefix = '@@INSIGHTS-CORE/NOTIFICATIONS/';
+export const ADD_NOTIFICATION = `${notificationsPrefix}ADD_NOTOFICATION`;
```
Comment (karelhala) [inline src/redux/action-types.js]: Typo: `ADD_NOTOFICATION` -> `ADD_NOTIFICATION`, `REMOVE_NOTOFICATION` -> `REMOVE_NOTIFICATION`, `CLEAR_NOTOFICATION` -> `CLEAR_NOTIFICATION`

File: src/redux/reducers/notifications/notifications.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import {
+    ADD_NOTIFICATION,
+    REMOVE_NOTIFICATION,
+    CLEAR_NOTIFICATIONS
+} from '../../action-types';
+import { generateID } from '../../../functions/generateID';
+
+const addNotification = (notitications, { payload }) => [ ...notitications, { ...payload, id: generateID('notyfication') }];
```
Comment (karelhala) [inline src/redux/reducers/notifications/notifications.js]: Typo: `notitications` ->  `notifications`

File: src/redux/reducers/notifications/notifications.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const removeNotification = (notifications, { payload }) => {
+    const index = notifications.findIndex(({ id }) => id === payload);
```
Comment (karelhala) [inline src/redux/reducers/notifications/notifications.js]: Please wrap the payload with parseInt: `parseInt(payload, 10)`

Comment (Hyperkid123) [reply src/redux/reducers/notifications/notifications.js]: @karelhala but the id generated by `generateID` is string, isn't it?

Comment (karelhala) [reply src/redux/reducers/notifications/notifications.js]: Oh yeah, that's right. Keep as is.

**Comment by Hyperkid123** [comment]
ping @karelhala

**Comment by AllenBW** [comment]
I woulda thought this would have been a feature bump not a bug fix (3.21.0 or something) 🤔

**Comment by Hyperkid123** [comment]
> I woulda thought this would have been a feature bump not a bug fix (3.21.0 or something) 🤔

ugh i forgot to add the release tag 😭

**Comment by AllenBW** [comment]
😭 🤗 👍  S'ALL GOOD ❤️ 

just wasn't sure.... 

iaf will grab and use none the less!

**Pull request merged**